pilot/pkg/xds/internalgen.go

```diff
@@ -150,6 +163,106 @@ func (sg *InternalGen) Generate(proxy *model.Proxy, push *model.PushContext, w *
+	case TypeDebugConfigDump:
+		if len(w.ResourceNames) == 0 {
+			log.Infof("%s without ResourceName", TypeDebugConfigDump)
+			break
+		}
+		var err error
+		res, err = sg.debugConfigDump(w.ResourceNames[0])
+		if err != nil {
+			log.Infof("%s failed: %v", TypeDebugConfigDump, err)
+			break
+		}
+	}
+	return res
+}
+
+// isSidecar ad-hoc method to see if connection represents a sidecar
+func isSidecar(con *Connection) bool {
+	return con != nil &&
+		con.node != nil &&
+		con.node.Metadata != nil &&
+		con.node.Metadata.ProxyConfig != nil
+}
+
+func (sg *InternalGen) debugSyncz() []*any.Any {
+	res := []*any.Any{}
+
+	stypes := []string{
+		pilot_xds_v3.ListenerShortType,
+		pilot_xds_v3.RouteShortType,
+		pilot_xds_v3.EndpointShortType,
+		pilot_xds_v3.ClusterShortType,
+	}
+
+	sg.Server.adsClientsMutex.RLock()
+	for _, con := range sg.Server.adsClients {
+		con.mu.RLock()
+		// Skip "nodes" without metdata (they are probably istioctl queries!)
+		if isSidecar(con) {
+			xdsConfigs := []*status.PerXdsConfig{}
+			for _, stype := range stypes {
+				pxc := &status.PerXdsConfig{}
+				if watchedResource, ok := con.node.Active[stype]; ok {
+					pxc.Status = debugSyncStatus(watchedResource)
+				} else {
+					pxc.Status = status.ConfigStatus_NOT_SENT
+				}
+				switch stype {
+				case pilot_xds_v3.ListenerShortType:
+					pxc.PerXdsConfig = &status.PerXdsConfig_ListenerConfig{}
+				case pilot_xds_v3.RouteShortType:
+					pxc.PerXdsConfig = &status.PerXdsConfig_RouteConfig{}
+				case pilot_xds_v3.EndpointShortType:
+					pxc.PerXdsConfig = &status.PerXdsConfig_EndpointConfig{}
+				case pilot_xds_v3.ClusterShortType:
+					pxc.PerXdsConfig = &status.PerXdsConfig_ClusterConfig{}
+				}
+				xdsConfigs = append(xdsConfigs, pxc)
+			}
+			clientConfig := &status.ClientConfig{
+				Node: &core.Node{
+					Id: con.node.ID,
+				},
+				XdsConfig: xdsConfigs,
+			}
+			log.Infof("Trying to show status of %#v\n", clientConfig)
+			res = append(res, util.MessageToAny(clientConfig))
+		}
+		con.mu.RUnlock()
 	}
+	sg.Server.adsClientsMutex.RUnlock()
+
 	return res
 }
+
+func debugSyncStatus(wr *model.WatchedResource) status.ConfigStatus {
+	if wr.NonceSent == "" {
+		return status.ConfigStatus_NOT_SENT
+	}
+	if wr.NonceAcked == wr.NonceSent {
+		return status.ConfigStatus_SYNCED
+	}
+	return status.ConfigStatus_STALE
+}
+
+func (sg *InternalGen) debugConfigDump(proxyID string) ([]*any.Any, error) {
+	conn := sg.Server.getProxyConnection(proxyID)
+	if conn == nil {
+		// Return no configuration.  Legacy Istio 1.6 /debug/config_dump
+		// returned http.StatusNotFound, but we do not consider it an error.
+		return []*any.Any{}, fmt.Errorf("config dump could not find connection for proxyID %q", proxyID)
+	}
+
+	dump, err := sg.Server.configDump(conn)
+	if err != nil {
+		// Return no configuration.  Legacy Istio 1.6 /debug/config_dump
+		// returned http.StatusInternalServerError, but we do not consider it an error.
+		return []*any.Any{}, err
```

Comment:
nit: `[]*any.Any{}` -> nil

Reply:
@howardjohn I believe I have fixed this and addressed your other comments.